axum/src/routing/mod.rs

```diff
@@ -162,20 +117,82 @@ where
             Err(service) => Endpoint::Route(Route::new(service)),
         };
 
-        let mut node =
-            Arc::try_unwrap(Arc::clone(&self.node)).unwrap_or_else(|node| (*node).clone());
-        if let Err(err) = node.insert(path, id) {
-            self.panic_on_matchit_error(err);
+        self.routes
+            .try_insert(path, endpoint)
+            .unwrap_or_else(|err| panic!("{}", err));
+
+        self
+    }
+
+    #[doc = include_str!("../docs/routing/nest.md")]
+    pub fn nest(mut self, mut path: &str, router: Router<B>) -> Self {
+        fn make_full_path<'a>(nested_path: &'a str, path: &'a str) -> Cow<'a, str> {
+            if nested_path == "/" {
+                path.into()
+            } else if path == "/" {
+                nested_path.into()
+            } else if let Some(path) = path.strip_suffix('/') {
+                format!("{}{}", path, nested_path).into()
+            } else {
+                format!("{}{}", path, nested_path).into()
+            }
         }
-        self.node = Arc::new(node);
 
-        self.routes.insert(id, service);
+        if path.is_empty() {
+            // nesting at `""` and `"/"` should mean the same thing
+            path = "/";
+        }
+
+        if path.contains('*') {
+            panic!("Invalid route: nested routes cannot contain wildcards (*)");
+        }
+
+        let prefix = path;
+
+        let Router {
+            routes,
+            custom_fallbacks,
+            // We consider the router on the right a "sub router" and therefore we always discards
+            // its default fallback. Custom fallbacks will be nested.
```

Comment:
Maybe also worth trying to convert a layered default fallback into a custom fallback so it's not lost.

But have to think about the edge cases and what might be least surprising.